### PR TITLE
8343132: Remove temporary transitions from Virtual thread implementation

### PR DESCRIPTION
**Pull request by AlanBateman**
This is an update to the Virtual thread implementation that we'd like to integrate in advance of JEP 491.

The update removes the use of  "temporary transitions", basically cases where the thread identity switches to the carrier thread to do something in the context of the carrier while a virtual thread is mounted. These cases create complexity for JVMTI and observability tools. It has also attracted attention in the review of the JEP 491 implementation as the object monitor changes have to deal with the possibility of entering monitors while in this state. There are 3 usages changes:

1. In submitRunContinuation the submit to the scheduler is changed so that it executes in the context of a virtual thread for cases where one virtual thread unparks another. This requires pinning to prevent preemption during this sensitive operation. ForkJoinPool.poolSubmit is changed so that it uses the identity of the carrier. This change has no impact on the uses of lazySubmit or externalSubmit.
2. Timed-park. The current implementation schedules/cancels the timer task with the virtual thread mounted. This runs in the context of the carrier as any contention would infer with thread state, park blocker and the parking permit. The implementation is changed to schedule the timeout after unmounting, and to cancel before re-mounting. The downside of this is that it will scheduled later (maybe 200us later than before). We could capture the time and adjust but it doesn't seem worth it.
3. jdk.tracePinnedThreads. This is a diagnostic option for finding usages of thread locals in code executed by virtual threads. This is changed so use a thread local to detect reentrancy.

The changes means that notifyJvmtiHideFrames, its intrinsic, and the JVMTI "tmp VTMS_transition" bit go away.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8343132](https://bugs.openjdk.org/browse/JDK-8343132): Remove temporary transitions from Virtual thread implementation (**Enhancement** - P3)

### Reviewers
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**) Review applies to [c88ce3dd](https://git.openjdk.org/jdk/pull/21735/files/c88ce3ddee80e0e42bb69265d44be1b84a28b5c4)
 * [Patricio Chilano Mateo](https://openjdk.org/census#pchilanomate) (@pchilano - **Reviewer**) Review applies to [c88ce3dd](https://git.openjdk.org/jdk/pull/21735/files/c88ce3ddee80e0e42bb69265d44be1b84a28b5c4)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21735/head:pull/21735` \
`$ git checkout pull/21735`

Update a local copy of the PR: \
`$ git checkout pull/21735` \
`$ git pull https://git.openjdk.org/jdk.git pull/21735/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21735`

View PR using the GUI difftool: \
`$ git pr show -t 21735`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21735.diff">https://git.openjdk.org/jdk/pull/21735.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21735#issuecomment-2443317266)
</details>

### CODE REVIEW

**Review by sspitsyn — APPROVED** [review]
The fix looks good to me. It is important and nice simplification.
The JVMTI part is strait forward.

**Review by pchilano — APPROVED** [review]
Looks good to me. Thanks for removing this.

File: src/java.base/share/classes/java/lang/ThreadLocal.java [hunk 1]
```diff
@@ -811,30 +810,28 @@ private static boolean traceVirtualThreadLocals() {
     }
 
     /**
-     * Print a stack trace if the current thread is a virtual thread.
+     * Print the print stack of the current thread, skipping the printStackTrace frame.
```
Comment (dholmes-ora) [inline src/java.base/share/classes/java/lang/ThreadLocal.java:813]:
```suggestion
     * Print the stack of the current thread, skipping the printStackTrace frame.
```

File: src/java.base/share/classes/java/lang/VirtualThread.java [hunk 2]
```diff
@@ -546,17 +527,20 @@ private void afterYield() {
...
-            int newState = (s == PARKING) ? PARKED : TIMED_PARKED;
-            setState(newState);
+            int newState;
+            if (s == PARKING) {
+                setState(newState = PARKED);
+            } else {
+                // schedule unpark
+                assert parkTimeout > 0;
+                timeoutTask = schedule(this::unpark, parkTimeout, NANOSECONDS);
+                setState(newState = TIMED_PARKED);
```
Comment (dholmes-ora) [inline src/java.base/share/classes/java/lang/VirtualThread.java:537]: Just to be clear here, if the timeout expires before we can call `setState`, the unpark is basically a no-op, and we will see that we have been unparked at line 541 and set the state correctly to UNPARKED.

File: test/jdk/java/lang/Thread/virtual/ParkWithFixedThreadPool.java [hunk 3]
```diff
@@ -74,4 +76,27 @@ public void run() {
...
+            pool = Executors.newFixedThreadPool(poolSize);
+        }
+
+        @Override
+        public void execute(Runnable task) {
+            try {
+                pool.execute(task);
+            } finally {
+                // ExecutorService::execute may consume parking permit
+                LockSupport.unpark(Thread.currentThread());
```
Comment (dholmes-ora) [inline test/jdk/java/lang/Thread/virtual/ParkWithFixedThreadPool.java:93]: This seems a bit odd - why would the current thread need to unpark itself? Why should it have a park permit available here?

**Review by dholmes-ora — APPROVED** [review]
Hotspot cleanup looks great! It is really good to see this temporary transition logic go away.

File: test/jdk/java/lang/Thread/virtual/ParkWithFixedThreadPool.java (hunk 3 above)
Comment (AlanBateman) [reply test/jdk/java/lang/Thread/virtual/ParkWithFixedThreadPool.java:93]: In this test, Scheduler.execute method will consume the current thread's parking permit when there is contention on the queue. In a well behaved system, all usages of park will first test some condition before parking. This test doesn't do this, hence it created the scenario where parking after unparking might hang. Previous discussion in [loom/pull/59](https://github.com/openjdk/loom/pull/59). There is no support exposed for doing custom schedulers at this time but this is the type of thing that comes up so we kept the test.

File: src/java.base/share/classes/java/lang/VirtualThread.java (hunk 2 above)
Comment (AlanBateman) [reply src/java.base/share/classes/java/lang/VirtualThread.java:537]: Yes, and same thing is unparked by some other thread while the target thread is parking. We have several tests that bash on this.

**Review by dholmes-ora — APPROVED** [review]

**Pull request closed without merging**